authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,134 @@
+            # obtain the user DN to bind with.
+            result_type, result_data = conn.result(ldap_result_id, 1)
+            if result_data:
+                if result_type != ldap.RES_SEARCH_RESULT:
+                    raise RuntimeError('An unexpected result object was '
+                                       'returned by the search: {result_type}'
+                                       .format(result_type=result_type))
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}' \
+                .format(e)
+            self.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
+
+        if len(result_data) != 1:
+            err = 'Expecting a single result matching the user with ' \
+                  'username {0}. Found {1}.' \
+                .format(who, len(result_data))
+            self.handle_authentication_error(err, conn)
+
+        # extract the user DN from the search result.
+        user_dn = result_data[0][0]
+
+        # attempt to bind using the matching user ID/DN and provided password.
+        self.validate_bind(user_dn, password)
+
+    def handle_authentication_error(self, err, conn):
+        conn.unbind()
+        self.lgr.warning(err)
+        raise AuthenticationException(err)
+
+    def validate_bind(self, who, password):
+            conn = self.create_bind_conn(who, password)
+            conn.unbind()
+
+    def create_bind_conn(self, who, password):
         # initialize connection to the LDAP server
         try:
-            conn = ldap.initialize(self.directory_url)
-        except Exception as e:
-            raise Exception(
-                'Failed to initialize LDAP connection to {0}; {1}'
-                .format(self.directory_url, str(e)))
+            conn = ldap.initialize(self.ldap_url)
+        except ldap.LDAPError as e:
+            err = 'Failed to initialize LDAP connection to {0}; {1}'\
+                .format(self.ldap_url, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
 
-        # trying to bind with the given user and password
+        # bind using the provided credentials
         try:
-            conn.bind_s(username, password)
-            return username
-        except Exception as e:
-            raise Exception(
-                'Failed to authenticate user {0}; {1}'
-                .format(username, str(e)))
-        finally:
-            conn.unbind_s()
+            conn.bind_s(who, password)
+        except ldap.INVALID_CREDENTIALS as e:
+            err = 'Invalid credentials provided for user: \'{0}\''\
+                .format(who)
+            self.handle_authentication_error(err, conn)
+        except ldap.LDAPError as e:
+            err = 'Failed to bind user \'{0}\'; {1}' \
+                .format(who, e)
+            self.handle_authentication_error(err, conn)
+
+        return conn
+
+    def is_search_based_authentication(self):
```

Comment:
1. we need identifying_attribute as well here
2. this is the place to check if some of the 4 mandatory settings are missing or None. if so, we should raise a value error with a detailed message